File: src/main/java/club/xiaojiawei/listener/log/ScreenLogListener.java
package club.xiaojiawei.listener.log;

import club.xiaojiawei.core.Core;
import club.xiaojiawei.custom.LogRunnable;
import club.xiaojiawei.data.SpringData;
import club.xiaojiawei.enums.ModeEnum;
import club.xiaojiawei.status.Mode;
import club.xiaojiawei.utils.SystemUtil;
import jakarta.annotation.Resource;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import org.apache.logging.log4j.util.Strings;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.io.*;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

/**
 * 游戏界面监听器
 * @author 肖嘉威
 * @date 2023/7/5 14:55
 */

@Slf4j
@Component
public class ScreenLogListener extends AbstractLogListener{

    @Resource
    private Core core;

    @Autowired
    public ScreenLogListener(SpringData springData) {
        super(springData.getScreenLogName(), 0, 1500, TimeUnit.MILLISECONDS);
    }

    @Override
    protected void readOldLog() throws IOException {
        String line;
        int index;
        ModeEnum finalMode = null;
        while ((line = accessFile.readLine()) != null) {
            if ((index = line.indexOf("currMode")) != -1) {
                finalMode = ModeEnum.valueOf(line.substring(index + 9));
            }
        }
        Mode.setCurrMode(finalMode);
    }

    @Override
    protected void listenLog() throws IOException, InterruptedException {
        String line;
        while (!isPause.get().get() && Strings.isNotBlank((line = accessFile.readLine()))){
            Mode.setCurrMode(resolveLog(line));
        }
        if (isPause.get().get()){
            cancelListener();
        }
    }

    @Override
    protected void otherListen() {

    }

    @Override
    protected void cancelOtherListener() {

    }

    private ModeEnum resolveLog(String line) throws InterruptedException {
        if (line == null){
            return null;
        }
        int index;
        if ((index = line.indexOf("currMode")) != -1){
            return ModeEnum.valueOf(line.substring(index + 9));
        }else if (line.contains("OnDestroy()")){
            Thread.sleep(2000);
            if (!SystemUtil.isAliveOfGame()){
                log.info("检测到游戏关闭，准备重启游戏");
                core.restart();
            }
        }
        return null;
    }
}

File: Hearthstone-Script/src/main/java/club/xiaojiawei/hsscript/listener/log/ScreenLogListener.kt
package club.xiaojiawei.hsscript.listener.log

import club.xiaojiawei.config.log
import club.xiaojiawei.enums.ModeEnum
import club.xiaojiawei.hsscript.consts.GAME_MODE_LOG_NAME
import club.xiaojiawei.hsscript.core.Core
import club.xiaojiawei.hsscript.listener.WorkTimeListener
import club.xiaojiawei.hsscript.status.Mode
import club.xiaojiawei.hsscript.status.PauseStatus
import club.xiaojiawei.hsscript.utils.GameUtil
import club.xiaojiawei.util.isFalse
import java.time.Duration
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.util.concurrent.TimeUnit

/**
 * 游戏界面监听器
 * @author 肖嘉威
 * @date 2023/7/5 14:55
 */
object ScreenLogListener :
    AbstractLogListener(GAME_MODE_LOG_NAME, 0, 50L, TimeUnit.MILLISECONDS) {

    private const val CURR_MODE_STR = "currMode="

    private const val CURR_MODE_STR_LEN = CURR_MODE_STR.length

    private const val NEXT_MODE_STR = "nextMode="

    private const val NEXT_MODE_STR_LEN = NEXT_MODE_STR.length

    private val formatter = DateTimeFormatter.ofPattern("HH:mm:ss.SSSSSSS")

    override fun dealOldLog() {
        var line: String
        var index: Int
        var finalCurrMode: ModeEnum? = null
        var finalNextMode: ModeEnum? = null
        while ((innerLogFile!!.readLine().also { line = it }) != null) {
            if ((line.indexOf(CURR_MODE_STR).also { index = it }) != -1) {
                finalCurrMode = ModeEnum.fromString(line.substring(index + CURR_MODE_STR_LEN))
                finalNextMode = null
            } else if ((line.indexOf(NEXT_MODE_STR).also { index = it }) != -1) {
                val blankIndex = line.indexOf(" ", index)
                finalNextMode = ModeEnum.fromString(line.substring(index + NEXT_MODE_STR_LEN, blankIndex))
                finalCurrMode = null
            }
        }
        finalCurrMode?.let {
            Mode.currMode = finalCurrMode
        } ?: let {
            finalNextMode?.let {
                Mode.nextMode = it
            }
        }

    }

    private var dealing = false

    override fun dealNewLog() {
        if (dealing) return
        dealing = true
        innerLogFile?.let {
            var line: String?
            while (!PauseStatus.isPause && WorkTimeListener.working) {
                line = it.readLine()
                if (line.isNullOrEmpty()) {
                    break
                }
                resolveLog(line)
            }
        }
        dealing = false
    }

    private fun resolveLog(line: String?) {
        line?.let { l ->
            var index: Int
            if ((l.indexOf(CURR_MODE_STR).also { index = it }) != -1) {
                val logTime = LocalTime.parse(line.substring(2, 18), formatter)
                val nowTime = LocalTime.now()
                val logDiffTime =
                    Duration.between(logTime, nowTime).toMillis()
                if (logDiffTime > 1500) {
                    log.warn { "${GAME_MODE_LOG_NAME}日志实际打印时间与输出时间相差过大，diff:${logDiffTime}，log:${line}，logTime:${logTime}，nowTime:${nowTime}" }
                }
                Mode.currMode = ModeEnum.fromString(l.substring(index + CURR_MODE_STR_LEN))
            } else if ((l.indexOf(NEXT_MODE_STR).also { index = it }) != -1) {
                val blankIndex = l.indexOf(" ", index)
                Mode.nextMode = ModeEnum.fromString(l.substring(index + NEXT_MODE_STR_LEN, blankIndex))
            } else if (l.contains("OnDestroy()")) {
                Thread.sleep(2000)
                GameUtil.isAliveOfGame().isFalse {
                    log.info { "检测到游戏关闭，准备重启游戏" }
                    Core.restart()
                }
            }
        }
    }
}
